communicator/ssh/communicator.go

```diff
@@ -585,7 +604,16 @@ func (c *comm) scpUploadSession(path string, input io.Reader, fi *os.FileInfo) e
 		return err
 	}
 	cmd.Wait()
+	if stdout.Len() > 0 {
+		return fmt.Errorf("%s", stdout.Bytes())
+	}
+	if stderr.Len() > 0 {
+		return fmt.Errorf("%s", stderr.Bytes())
+	}
 	if cmd.ExitStatus == 0 {
+		if fi == nil {
+			return fmt.Errorf("Upload path is a directory, unable to continue.")
+		}
```

Comment:
My brain is really really sluggish today. When is fi nil?

Reply:
more details in https://github.com/hashicorp/packer/issues/6032